bigtable/benchmarks/setup.cc

```diff
@@ -0,0 +1,110 @@
+// Copyright 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "bigtable/benchmarks/setup.h"
+#include <absl/strings/str_cat.h>
+#include <cctype>
+#include <iomanip>
+#include <sstream>
+#include "bigtable/benchmarks/random.h"
+#include "bigtable/client/build_info.h"
+
+/// Supporting types and functions to implement `BenchmarkSetup`
+namespace {
+std::string FormattedStartTime() {
+  auto start = std::chrono::system_clock::now();
+  std::time_t start_c = std::chrono::system_clock::to_time_t(start);
+  std::string formatted("YYYY-MM-DDTHH:SS:MMZ");
+  std::strftime(&formatted[0], formatted.size(), "%FT%TZ",
+                std::gmtime(&start_c));
+  return formatted;
+}
+
+std::string FormattedAnnotations() {
+  std::string notes =
+      absl::StrCat(bigtable::version_string(), ";", bigtable::compiler, ";",
+                   bigtable::compiler_flags);
+  std::transform(notes.begin(), notes.end(), notes.begin(),
+                 [](char c) { return c == '\n' ? ';' : c; });
+  return notes;
+}
+
+std::string MakeRandomTableId(std::string const& prefix) {
+  static std::string const table_id_chars(
+      "ABCDEFGHIJLKMNOPQRSTUVWXYZabcdefghijlkmnopqrstuvwxyz0123456789_");
+  auto gen = bigtable::benchmarks::MakeDefaultPRNG();
+  return absl::StrCat(
+      prefix, "-",
+      bigtable::benchmarks::Sample(
+          gen, bigtable::benchmarks::kTableIdRandomLetters, table_id_chars));
+}
+}  // anonymous namespace
+
+namespace bigtable {
+namespace benchmarks {
+BenchmarkSetup::BenchmarkSetup(std::string const& prefix, int& argc,
+                               char* argv[])
+    : start_time_(FormattedStartTime()),
+      notes_(FormattedAnnotations()),
+      project_id_(),
+      instance_id_(),
+      table_id_(MakeRandomTableId(prefix)) {
+  if (argc < 3) {
```

Comment:
throw otherwise?

Reply:
Yes, it throws if there are not enough arguments.  This is a constructor, there is no other way to signal errors.  The intention is for `main()` to capture the exception exit with 1.  Did I understand the question?

Reply:
Yes, sorry. All makes sense.